### PR TITLE
fix(ops): harden operator review idempotency

### PR DESCRIPTION
**Pull request by mt4110**
## Design source

`ISSUE-12-operator-review-appeal-evidence.md`

This is design ISSUE-12, not a GitHub issue number. The GitHub issue number is intentionally not hardcoded.

## Summary

Follow-up hardening for the operator review / appeal / evidence workflow merged in PR #25.

This PR keeps the ISSUE-12 boundary intact: operator decisions remain append-only facts, Promise / settlement writer truth is not overwritten, user-facing reason codes stay bounded, raw evidence locators and internal operator notes do not leak into user-facing read models, and ISSUE-13 / ISSUE-14 remain deferred consumers.

## Current commit stack

Base hardening commit:
- `eed1bf5` `fix(ops): harden operator review idempotency`

Latest review-feedback commits:
- `c6a8fa3` `docs(ops): clarify operator review replay hash comments`
- `469e9ad` `fix(ops): compare operator review replays against persisted hashes`
- `cf7b4fc` `fix(ops): normalize legacy replay realm ids`
- `a2e54b4` `fix(ops): backfill legacy operator review replay hashes`
- `e122b36` `fix(ops): canonicalize operator review idempotency hashes`

## What changed

- Added forward-only migration `0015_operator_review_hardening.sql`
- Added payload hash columns for review case, operator decision, and appeal idempotency replay checks
- Canonicalized replay hash inputs by recursively sorting JSON object keys, preserving array order, and using stable JSON escaping for strings and object keys
- Changed normal replay mismatch checks to compare persisted hashes instead of reloading source snapshots, internal operator notes, decision payloads, appellant statements, or new evidence summaries
- Added legacy `NULL` hash handling that recomputes the hash from preserved rows once, backfills the derived hash column, and then compares against the persisted hash
- Normalized legacy `related_realm_id` hashing to match the forward request path
- Ensured concurrent backfill races compare against the value actually persisted in PostgreSQL
- Removed remaining `SELECT *` / `RETURNING *` reads from the operator review repository path
- Added edge-case tests for review case / decision / appeal idempotency drift, canonical JSON key ordering, legacy hash backfill, and concurrent replay safety
- Updated ISSUE-12 docs and migration comments to clarify the replay-hash boundary

## Legacy backfill policy

- New rows write 64-character lowercase SHA-256 replay payload hashes at creation time.
- Normal replays compare the incoming canonical payload hash against the persisted hash.
- Legacy rows with `NULL` replay hashes are not blindly accepted. The handler reconstructs the hash from the preserved writer row, backfills the derived hash with a guarded `UPDATE ... WHERE ... IS NULL RETURNING`, falls back to reading the persisted hash if another request won the race, and then performs the same hash comparison as normal replay.
- This legacy fallback is limited to the idempotency replay path and does not change authoritative writer truth.

## Core behavior preserved

- Operator decisions remain append-only facts
- Operator decisions do not mutate Promise / settlement writer truth
- User-facing reason codes remain bounded
- Raw evidence locators and internal operator notes do not leak into user-facing read models
- Evidence access remains scoped, auditable, and role-guarded
- ISSUE-13 and ISSUE-14 remain deferred consumers

## Validation

- `make ENV_FILE=.env.example db-bootstrap`
- `make ENV_FILE=.env.example db-migrate`
- `make ENV_FILE=.env.example db-status`
- `cargo check`
- `cargo test --test operator_review --no-fail-fast`
- `cargo test --no-fail-fast`
- `git diff --check`

Note: `cargo fmt --check` still reports the pre-existing formatting-only preference in `apps/backend/src/handlers/orchestration.rs`. This PR leaves that unrelated file untouched.

## Deferred

- ISSUE-13 room progression state machine
- ISSUE-14 Promise UI / proof missing fallback UI
- settlement UI
- full dispute center UI
- proof persistence
- broad operator console product work

### CODE REVIEW

File: apps/backend/src/services/operator_review/repository.rs [hunk 1]
```diff
@@ -1420,30 +1528,72 @@ async fn sync_review_case_status_tx<C: GenericClient + Sync>(
...
+        "new_evidence_summary_json": &input.new_evidence_summary_json,
+    }))
+}
+
+fn optional_uuid_hash_value(value: &Option<Uuid>) -> Option<String> {
+    value.map(|value| value.to_string())
+}
+
+fn hash_json_value(value: &Value) -> String {
+    let digest = Sha256::digest(value.to_string().as_bytes());
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/operator_review/repository.rs]: 対応済みです。canonical hash + legacy backfill 対応済み。

- `e122b36` で JSON object key を再帰 sort する canonical hash に変更しました。
- `a2e54b4` で legacy `NULL` replay hash の backfill path を追加しました。
- canonical key-order replay の regression test も追加済みです。

File: apps/backend/src/services/operator_review/repository.rs [hunk 2]
```diff
@@ -1420,30 +1528,72 @@ async fn sync_review_case_status_tx<C: GenericClient + Sync>(
...
-        || existing_related_realm_id != normalize_optional(&input.related_realm_id)
-        || row.get::<_, String>("opened_reason_code") != input.opened_reason_code
-        || row.get::<_, String>("source_fact_kind") != input.source_fact_kind
-        || row.get::<_, String>("source_fact_id") != input.source_fact_id
-        || row.get::<_, Value>("source_snapshot_json") != input.source_snapshot_json
-        || row.get::<_, Option<Uuid>>("assigned_operator_id") != *assigned_operator_id
-    {
+    let existing_hash: Option<String> = row.get("request_payload_hash");
+    if existing_hash.as_deref() != Some(request_payload_hash) {
         return Err(OperatorReviewError::BadRequest(
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/operator_review/repository.rs]: 対応済みです。canonical hash + legacy backfill 対応済み。

- `a2e54b4` で review case / decision / appeal の legacy `NULL` replay hash を preserved row から backfill する path を追加しました。
- backfill 後は通常 replay と同じ persisted hash 比較に戻します。
- 既存 writer truth は書き換えず、derived replay hash column だけを補完します。

File: apps/backend/src/services/operator_review/repository.rs [hunk 3]
```diff
@@ -1469,23 +1674,182 @@ fn ensure_operator_decision_matches_input(
...
+    let assigned_operator_id: Option<Uuid> = row.get("assigned_operator_id");
+
+    hash_json_value(&json!({
+        "schema_version": 1,
+        "case_type": row.get::<_, String>("case_type"),
+        "severity": row.get::<_, String>("severity"),
+        "subject_account_id": optional_uuid_hash_value(&subject_account_id),
+        "related_promise_intent_id": optional_uuid_hash_value(&related_promise_intent_id),
+        "related_settlement_case_id": optional_uuid_hash_value(&related_settlement_case_id),
+        "related_realm_id": row.get::<_, Option<String>>("related_realm_id"),
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/operator_review/repository.rs]: 対応済みです。canonical hash + legacy backfill 対応済み。

`cf7b4fc` で legacy backfill 時の `related_realm_id` を forward request path と同じく normalize するようにしました。padded / whitespace-only realm id の replay regression も追加済みです。

File: apps/backend/src/services/operator_review/repository.rs [hunk 4]
```diff
@@ -1469,23 +1674,182 @@ fn ensure_operator_decision_matches_input(
...
+        "opened_reason_code": row.get::<_, String>("opened_reason_code"),
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/operator_review/repository.rs:1773]: 対応済みです。canonical hash + legacy backfill 対応済み。

`cf7b4fc` で `review_case_payload_hash_from_row` 側も `related_realm_id` を normalize してから hash input に入れるようにしました。

File: apps/backend/src/services/operator_review/repository.rs [hunk 5]
```diff
@@ -1469,23 +1674,183 @@ fn ensure_operator_decision_matches_input(
...
+            UPDATE dao.review_cases
+            SET request_payload_hash = $2
+            WHERE review_case_id = $1
+              AND request_payload_hash IS NULL
+            ",
+            &[&review_case_id, &payload_hash],
+        )
+        .await
+        .map_err(db_error)?;
+    Ok(payload_hash)
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/operator_review/repository.rs]: 対応済みです。canonical hash + legacy backfill 対応済み。

`469e9ad` で guarded backfill の `UPDATE ... IS NULL RETURNING` を使い、別 request が先に backfill した場合は persisted hash を読み直して比較するようにしました。caller は計算値ではなく PostgreSQL に保存された hash と比較します。

File: apps/backend/src/services/operator_review/repository.rs [hunk 6]
```diff
@@ -1420,30 +1529,125 @@ async fn sync_review_case_status_tx<C: GenericClient + Sync>(
...
+    value.map(|value| value.to_string())
+}
+
+fn hash_json_value(value: &Value) -> String {
+    let digest = Sha256::digest(canonical_json_text(value).as_bytes());
+    let mut encoded = String::with_capacity(digest.len() * 2);
+    for byte in digest {
+        let _ = write!(&mut encoded, "{byte:02x}");
+    }
+    encoded
```
[the remark this answers is not in the document]
Comment (mt4110) [reply apps/backend/src/services/operator_review/repository.rs:1588]: 確認しました。canonical hash + legacy backfill 対応済みです。

今回は ISSUE-12 baseline の replay payload サイズと実装の単純さを優先し、canonical JSON string を作ってから SHA-256 する形に留めています。必要になれば streaming hasher に差し替え可能な局所 helper に閉じています。
